### PR TITLE
Kubevirt VmDetails

### PR DESCRIPTION
**Pull request by mareklibra**
Introduces Details Page for Virtual Machines.

Depends on: 
- [x] #1592 
- [x] #1711 
- [x] #1725
- [x] #1738 

---
01_vmDetail

---
02_vmDetail

### CODE REVIEW

File: frontend/packages/_plugin-style.scss [hunk 1]
```diff
@@ -0,0 +1,4 @@
+/* Styles of all active plugins */
```
Comment (mareklibra) [inline frontend/packages/_plugin-style.scss]: Simple, but working.
See additional comments for more details.

**Comment by mareklibra** [comment]
To achieve full encapsulation of plugins, styles should be extendable as well.
We can handle that on plugin's package.json level with modification to webpack.
@vojtechszocs , wdyt?

If such an extension point is considered as waste of effort, we can go with the simple solution proposed here:
- if a plugin needs custom styling, there will be single entry `src/style.scss` in its root
- `frontend/packages/_plugin-style.scss` is manually maintained to register all styling of active plugins

So to (de)activate a plugin, one-line modification of `frontend/packages/_plugin-style.scss` and `console-app/package.json` is required

**Comment by mareklibra** [comment]
@spadgett , can I ask you for review here, please?

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
+import * as React from 'react';
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: Can we move this to `./vm/` folder? - to distinguish between vm and vm template files in the future

File: frontend/packages/kubevirt-plugin/src/components/constants.js [hunk 3]
```diff
@@ -0,0 +1 @@
+export const DASHES = '---';
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/constants.js]: this should probably go to `console-shared`

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+// import VmConsolesConnected from '../vmconsoles';
+// import { Nic } from '../nic';
+// import { Disk } from '../disk';
+// import { menuActions } from './menu-actions';
+
+export const VmDetailsFirehose = ({ obj: vm, ...rest }: { obj: VmKind }) => {
+  const { name, namespace } = vm.metadata;
+
+  const resources = [
+    getResource(VirtualMachineInstanceModel, { name, namespace, isList: false }),
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: we could probably move this method here rather sooner than later, because it is used in many places.

- note: it builds requested resource object with reasonable defaults

File: frontend/packages/kubevirt-plugin/src/utils/resources.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+import * as _ from 'lodash-es';
+
+export const getLoadedResource = (props, kind) =>
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/utils/resources.ts]: This kind of functionality should be probably included in the firehose. 
Many components do not care about the loaded prop. They just need the data or undefined/null if the data is not available yet.

- note: previously implemented by WithResources.

Comment (spadgett) [reply frontend/packages/kubevirt-plugin/src/utils/resources.ts]: Components can choose to handle `!loaded` the same as no data, but that should be up to the component. It's a worse API not to have an explicit `loaded` prop. In most cases, you should handle it as well as handle errors.

File: frontend/packages/kubevirt-plugin/src/components/constants.js (hunk 3 above)
Comment (spadgett) [reply frontend/packages/kubevirt-plugin/src/components/constants.js]: We use a single dash, `-`, throughout console. Is there a reason it's `---` here?

File: frontend/packages/kubevirt-plugin/src/utils/resources.ts (hunk 5 above)
Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/utils/resources.ts]: I am not suggesting to make any regression to the API. It is enough Just to enhance it, possibly by adding some flags to the resource (or the whole firehose).

Many components need just the raw data and this would make it easier for them to have less logic or a need to specify getLoadedResource for each resource (like this PR).

Comment (spadgett) [reply frontend/packages/kubevirt-plugin/src/utils/resources.ts]: In most cases, components *should* explicitly handle `loaded`. Changing the API to make it easy to ignore seems like a bad change.

I also think `getLoadedResource` is a not a useful helper because it assumes the property of the data. The prop is not always the kind. It would be better to remove it from the PR.

File: frontend/packages/kubevirt-plugin/src/components/constants.js (hunk 3 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/constants.js]: IIRC, this was requested during UX review some time ago.
But I agree, it makes sense to unify and potentially change later everywhere.

File: frontend/packages/kubevirt-plugin/src/utils/resources.ts (hunk 5 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/utils/resources.ts]: Recently I try to avoid `WithResources` in favour of just `Firehose`. There will be probably a small change to `Firehose` needed when migrating later features, but it's not a recent concern.

The purpose of this `getLoadedResource` function is to provide simple mapping to meet contract with existing component. So far, similar logic was handled via `WithResources` which took care of error handling and potentially passed empty data to the component or tentatively rendered a loader component.

I can improve components in `kubevirt-web-ui-components` to deal with the `Firehose` state machine explicitly. But it's extra complexity as I can't break the "safe-path" for _potentially_  releasing web-ui even in 4.2 timeframe. 

I will try, but definitely I'll definitely need to create new release of `web-ui-components` for that.

File: frontend/packages/_plugin-style.scss (hunk 1 above)
Comment (mareklibra) [reply frontend/packages/_plugin-style.scss]: Can I ask for comments here, please? @spadgett , @vojtechszocs

File: frontend/packages/kubevirt-plugin/src/components/constants.js (hunk 3 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/constants.js]: One more change: the `DASHES` are imported from `kubevirt-web-ui-components` now, so triple dash `---` is still used per former UXD review. This way it will be consistent across Virtual Machines pages and can be unified with OpenShift when moving `kubevirt-web-ui-components` under `console`. Or if it is a blocker, I can fix there and fire new release. @spadgett , are you ok with that, please?

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx (hunk 4 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: And it appeared that it can be even farther enhanced. As discussed before, it will be moved along the rest of `kubevirt-web-ui-components`.

File: frontend/packages/kubevirt-plugin/src/utils/resources.ts (hunk 5 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/utils/resources.ts]: When looking into the code once again, I think it's better to enhance `Firehose` (see #1711) and we can move checks to children components consistently with the rest of `console`. Please have a look at recent changes in this PR.

File: frontend/packages/kubevirt-plugin/src/components/constants.js (hunk 3 above)
Comment (spadgett) [reply frontend/packages/kubevirt-plugin/src/components/constants.js]: Sorry, I disagree. We should change it everywhere in console or not at all. It sounds like we reviewed one page and didn't consider any of the existing pages that already have a dash. IMO, the right thing to do is fix kubevrit-web-ui-components so that it's consistent with existing console pages.

We're adding kubevirt as a plugin is so that it feels like an integrated part of console. This might seem like a little thing, but it undercuts that goal. The kubevirt pages should look and behave like other console pages.

For the record, I prefer a single dash `-`. But I'm most concerned about keeping consistency.

**Comment by spadgett** [comment]
This page does not look or behave like other console pages. This goes against the goal of making kubevirt feel like an integrated, first-class part of console.

* It doesn't use the standard `ResourceSummary` component on every other detail page
* It uses more columns than every other details page
* Empty values should be a single `-` with `text-muted` style
* The `Edit` buttons should be a menu item in an actions dropdown

What does the floating (?) icon do?

**Comment by lizsurette** [comment]
> This page does not look or behave like other console pages. This goes against the goal of making kubevirt feel like an integrated, first-class part of console.

@mareklibra Here is our updated design based on feedback from @beanh66 and the OpenShift design team...


You can track the larger [design PR](https://github.com/openshift/openshift-origin-design/pull/200) if needed, too.

FYI @matthewcarleton

File: frontend/packages/_plugin-style.scss (hunk 1 above)
Comment (vojtechszocs) [reply frontend/packages/_plugin-style.scss]: @mareklibra Sorry for responding late.

Plugin-specific styles can be loaded as a side effect of loading the plugin's entry module, e.g. in `kubevirt-plugin/src/plugin.ts`:

```ts
import './style.scss';
```

I see you've already updated the PR to do that, so :+1: on that.

Comment (vojtechszocs) [reply frontend/packages/_plugin-style.scss]: @mareklibra Please note, once #1708 is merged and you enable the `kubevirt-plugin` & run tests (`yarn test`), we'll need to ensure Jest properly mocks `scss` files. Right now, the relevant module mapper config is:

```
"\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js"
```

So we'll have to add `scss` there ^^.

**Comment by spadgett** [comment]
Thanks @lizsurette. Looks good!

Note that `ResourceSummary` forces a certain layout of the first few fields for consistency across pages. So the order of the items in the left column might change slightly. (Namespace is always directly under name for instance.)

**Comment by lizsurette** [comment]
> Note that `ResourceSummary` forces a certain layout of the first few fields for consistency across pages. So the order of the items in the left column might change slightly. (Namespace is always directly under name for instance.)

That makes sense to me @spadgett. FYI @matthewcarleton

**Comment by matthewcarleton** [comment]
I've suggested on the updated design PR that we include designs around the inline edit rather than just omitting the bulk for clarity. @spadgett @lizsurette

**Comment by mareklibra** [comment]
@lizsurette , @spadgett , thanks for the review.

Recent design conforms web-ui. Original plan was to move our fork to upstream as it is with just minor necessary modifications. Improvements were supposed to be done in follow-ups to not delay the effort more then necessary while targeting full feature parity for 4.2.

I agree that the pages should look and behave consistently and especially changes on upstream side should be reflected here.
I will adapt this PR accordingly. Anyway, I suggest to implement new features (like the charts) still in follow-ups.

@jelkosz, FYI., this will take a little bit more time then originally expected.

**Comment by jelkosz** [comment]
> I will adapt this PR accordingly. Anyway, I suggest to implement new features (like the charts) still in follow-ups.

this makes sense to me! Redesign to look like the rest of the pages but without adding new features is a good approach.

> FYI., this will take a little bit more time then originally expected.

:(

**Comment by vojtechszocs** [comment]
> To achieve full encapsulation of plugins, styles should be extendable as well.
> We can handle that on plugin's package.json level with modification to webpack.
> @vojtechszocs , wdyt?
> 
> If such an extension point is considered as waste of effort, we can go with the simple solution proposed here:
> 
> * if a plugin needs custom styling, there will be single entry `src/style.scss` in its root
> * `frontend/packages/_plugin-style.scss` is manually maintained to register all styling of active plugins
> 
> So to (de)activate a plugin, one-line modification of `frontend/packages/_plugin-style.scss` and `console-app/package.json` is required

Again, sorry for responding late. Based on discussion with @christianvogt I think the best approach at the moment is to `import` the necessary stylesheets within the plugin's entry module.

Enabling a plugin shouldn't involve more steps beyond updating `console-app`'s dependencies.

Ideally, plugin entry modules should be side-effect-free and only export the list of extensions, because they are loaded early and therefore impact the initial chunk size. We can revisit this in future and have a better separation.

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx (hunk 2 above)
Comment (vojtechszocs) [reply frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: :+1: `public/components/*` seem to use the plural for directories, so I'd suggest using `components/vms` and `components/vm-templates` for better organization.

File: frontend/packages/kubevirt-plugin/src/components/constants.js (hunk 3 above)
Comment (vojtechszocs) [reply frontend/packages/kubevirt-plugin/src/components/constants.js]: Can't we simply use the shared package here?

```ts
// packages/console-shared/src/constants/index.ts
export const DASH = '-';
```

File: frontend/packages/_plugin-style.scss (hunk 1 above)
Comment (vojtechszocs) [reply frontend/packages/_plugin-style.scss]: This is addressed by #1721 ([commit](https://github.com/openshift/console/pull/1721/commits/6db90c663cbdad13fb0e86b6d480c1829aab3ec5)).

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+      kind={PodModel.kind}
+      name={getName(pod)}
+      namespace={getNamespace(pod)}
+      uid={pod.metadata.uid}
+    />
+  ) : (
+    DASHES
+  );
+
+const VmDetailsConnected = (props: VmDetailsProps) => {
```
Comment (spadgett) [inline frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]:
```suggestion
const VMDetailsConnected: React.FC<VMDetailsProps> = (props) => {
```

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  const { vm } = props;
```
Comment (spadgett) [inline frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: Consider

```suggestion
  const { vm, ...rest } = props;
```

and using `{...rest}` below

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx (hunk 6 above)
Comment (spadgett) [reply frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: I'm confused: this is `VMDetailsConnected`, but I don't see a call to `connect()`. We should give this a different name if it's not referring to redux connect

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+export const VmDetailsFirehose = ({ obj: vm }: { obj: VmKind }) => {
+  const { name, namespace } = vm.metadata;
+
+  const vmiRes = getResource(VirtualMachineInstanceModel, {
+    name,
+    namespace,
+    isList: false,
+    props: 'vmi',
+  });
+  vmiRes.ignoreIfMissing = true;
```
Comment (spadgett) [inline frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: `optional`

Add this to declaration. No reason this needs to be on a separate line.

File: frontend/public/components/utils/firehose.jsx [hunk 9]
```diff
@@ -25,8 +25,12 @@ const processReduxId = ({k8s}, props) => {
   }
 
   if (!isList) {
-    const stuff = k8s.get(reduxID);
-    return stuff ? stuff.toJS() : {};
+    let stuff = k8s.get(reduxID);
+    if (stuff) {
+      stuff = stuff.toJS();
+      stuff.ignoreIfMissing = props.ignoreIfMissing;
```
Comment (spadgett) [inline frontend/public/components/utils/firehose.jsx]: This should be removed since #1711 merged

Comment (mareklibra) [reply frontend/public/components/utils/firehose.jsx]: resolved by rebase

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx (hunk 7 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: done

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx (hunk 6 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: done

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+  );
+};
+
+const podResourceLink = ({ pod }: { pod?: PodKind }) =>
+  pod ? (
+    <ResourceLink
+      kind={PodModel.kind}
+      name={getName(pod)}
+      namespace={getNamespace(pod)}
+      uid={pod.metadata.uid}
```
[the remark this answers is not in the document]
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: done

File: frontend/packages/kubevirt-plugin/src/types.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
+import { K8sResourceKind, ObjectMetadata } from '@console/internal/module/k8s';
+
+// https://kubevirt.io/api-reference/master/definitions.html#_v1_virtualmachineinstancespec
+export type VmiSpec = {
```
[the remark this answers is not in the document]
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/types.ts]: done, changed on multiple places

File: frontend/packages/kubevirt-plugin/src/components/vm-details.tsx [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    navFactory.details(VmDetailsFirehose),
+    navFactory.editYaml(),
+    // consolePage,
+    // navFactory.events(VmEvents),
+    // nicsPage,
+    // disksPage,
+  ];
+
+  const breadcrumbsFor = (obj) =>
+    breadcrumbsForOwnerRefs(obj).concat({
```
[the remark this answers is not in the document]
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vm-details.tsx]: Rebased on top of that and updated accordingly

File: frontend/packages/kubevirt-plugin/src/components/vms/utils.tsx [hunk 13]
```diff
@@ -0,0 +1,3 @@
+import * as React from 'react';
+
+export const Dashes = () => <span className="text-muted">-</span>;
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/vms/utils.tsx]: why is this necessary? Isn't just a string enough?

- nevertheless this should be exported as a component from `@console/shared`

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+export const VMDetailsFirehose = ({ obj: vm }: { obj: VMKind }) => {
+  const { name, namespace } = vm.metadata;
+
+  const vmiRes = getResource(VirtualMachineInstanceModel, {
+    name,
+    namespace,
+    isList: false,
+    prop: 'vmi',
+  });
+  vmiRes.optional = true;
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx]: nit: this could be handled by getResource

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+      <dd>{getDescription(vm)}</dd>
+      <dt>Operating System</dt>
+      <dd>{getOperatingSystemName(vm) || getOperatingSystem(vm) || <Dashes />}</dd>
+      <dt>Template</dt>
+      <dd>{templateLink || <Dashes />}</dd>
+    </ResourceSummary>
+  );
+};
+
+export const VMDetailsList = ({ vm, vmi, pods, migrations }: VMResourceListProps) => {
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx]: are you planning to do the editing part as well? IMO it would be much easier to pull it in atomically than by parts.

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+  const templateLink = template && getTemplateDisplayName(template); // TODO(mlibra): link to a template detail, once implemented
+
+  return (
+    <ResourceSummary resource={vm}>
+      <dt>Description</dt>
+      <dd>{getDescription(vm)}</dd>
+      <dt>Operating System</dt>
+      <dd>{getOperatingSystemName(vm) || getOperatingSystem(vm) || <Dashes />}</dd>
+      <dt>Template</dt>
+      <dd>{templateLink || <Dashes />}</dd>
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx]: nit: missing the warning about missing template

File: frontend/packages/kubevirt-plugin/src/components/vms/utils.tsx (hunk 13 above)
Comment (spadgett) [reply frontend/packages/kubevirt-plugin/src/components/vms/utils.tsx]: Sorry I might have been wrong on using `text-muted` here. We should do whatever is done elsewhere. (Hopefully we're consistent.)

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx (hunk 15 above)
Comment (jelkosz) [reply frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx]: Editing will have to wait for a followup patch, we need to get at least this part in.

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx (hunk 14 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx]: Right, this is the "enhancement" I wrote about somewhere.
I will put it there but let's not block this PR on that.

File: frontend/packages/kubevirt-plugin/src/components/vms/utils.tsx (hunk 13 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vms/utils.tsx]: ok, changed to use single dash now.

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx (hunk 16 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx]: Templates are missing in UI at all, they will be added in a follow-up along adding proper resource link here. Recently either name or a dash is rendered.

**Comment by mareklibra** [comment]
Rebased and addressed comments.

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx (hunk 15 above)
Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx]: ok

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+import { VirtualMachineInstanceModel, VirtualMachineInstanceMigrationModel } from '../../models';
+import { VMResourceSummary, VMDetailsList } from './vm-resource';
+
+// import { VmEvents } from './vm-events';
+// import VmConsolesConnected from '../vmconsoles';
+// import { Nic } from '../nic';
+// import { Disk } from '../disk';
+// import { menuActions } from './menu-actions';
+
+export const VMDetailsFirehose = ({ obj: vm }: { obj: VMKind }) => {
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx]: Can we put the VmDetails implementation into a separate file and use this file only as disambiguation for all the tabs?

Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx]: Considering its purpose, I think it's better to keep it collocated. I've removed unnecessary `export` - I'm sorry for confusion.

File: frontend/packages/kubevirt-plugin/src/components/vms/types.ts [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+export type VMStatus = {
+  conditions?: any[];
+  created?: boolean;
+  ready?: boolean;
+  stateChangeRequests?: any[];
+};
+
+// https://kubevirt.io/api-reference/master/definitions.html#_v1_virtualmachine
+export type VMKind = {
```
[the remark this answers is not in the document]
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vms/types.ts]: Honestly, I don't think this is something crucial. Anyway, the rest of recent code-base uses both options where how to place operands.

The one already used in this PR looks more common (statistic retrieved via `grep`).

**Comment by mareklibra** [comment]
@spadgett , can I ask you for review, please?

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx (hunk 17 above)
Comment (mareklibra) [reply frontend/packages/kubevirt-plugin/src/components/vms/vm-details.tsx]: Other pages have details co-located as well. Anyway, to move forward, I split it here.

**Comment by suomiy** [comment]
Redesigned to fit into openshift design. All review comments have been addressed + any other potential comments can be addressed in followups. 

Merging in order to unblock other PRs.

**Comment by mareklibra** [comment]
Last force-push was done by mistake while rebasing other PRs stacked on top of this. Should be ok now. I'm sorry for confusion.

File: frontend/packages/kubevirt-plugin/src/components/vms/types.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/packages/kubevirt-plugin/src/components/vms/types.ts]: > Honestly, I don't think this is something crucial.

IMO it's more readable to put base object before the extension, instead of the other way around. So my comment was related to better code readability.
